File: app/src/main/java/com/deadcells/modding/AtlasActivity.java
package com.deadcells.modding;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

public class AtlasActivity extends Activity {
    private TextView mStatus;
    private final Handler mHandler = new Handler(Looper.getMainLooper());

    private void startOp(Runnable r) {
        new Thread(r).start();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ScrollView rootScroll = new ScrollView(this);
        rootScroll.setFillViewport(true);
        rootScroll.setLayoutParams(new LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT));

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(48, 48, 48, 48);
        rootScroll.addView(layout);

        TextView title = new TextView(this);
        title.setText(getString(R.string.atlas_title));
        title.setTextSize(22);
        layout.addView(title);

        // --- Atlas unpack ---
        TextView ul = new TextView(this);
        ul.setText("\n" + getString(R.string.pak_atlas_unpack)); ul.setTextSize(16);
        layout.addView(ul);
        EditText ai = new EditText(this); ai.setHint(getString(R.string.pak_atlas_unpack_hint)); ai.setTextSize(14);
        layout.addView(ai);
        EditText ao = new EditText(this); ao.setHint(getString(R.string.pak_unpack_hint_dir)); ao.setTextSize(14);
        layout.addView(ao);
        Button ub = new Button(this); ub.setText(getString(R.string.pak_atlas_unpack_btn));
        ub.setOnClickListener(v -> {
            String i = ai.getText().toString().trim(), o = ao.getText().toString().trim();
            if (i.isEmpty() || o.isEmpty()) { Toast.makeText(this, getString(R.string.pak_fill_paths), Toast.LENGTH_SHORT).show(); return; }
            startOp(() -> {
                boolean ok = PakTool.atlasUnpack(i, o);
                mHandler.post(() -> mStatus.setText(ok ?
                    getString(R.string.pak_complete, o) : getString(R.string.pak_failed)));
            });
        });
        layout.addView(ub);

        // --- Atlas pack ---
        TextView pl = new TextView(this);
        pl.setText("\n" + getString(R.string.pak_atlas_pack)); pl.setTextSize(16);
        layout.addView(pl);
        EditText apd = new EditText(this); apd.setHint(getString(R.string.pak_atlas_pack_hint_dir)); apd.setTextSize(14);
        layout.addView(apd);
        EditText apa = new EditText(this); apa.setHint(getString(R.string.pak_atlas_pack_hint_atlas)); apa.setTextSize(14);
        layout.addView(apa);
        EditText app = new EditText(this); app.setHint(getString(R.string.pak_atlas_pack_hint_png)); app.setTextSize(14);
        layout.addView(app);
        Button pb = new Button(this); pb.setText(getString(R.string.pak_atlas_pack_btn));
        pb.setOnClickListener(v -> {
            String d = apd.getText().toString().trim(), a = apa.getText().toString().trim(), p = app.getText().toString().trim();
            if (d.isEmpty() || a.isEmpty() || p.isEmpty()) { Toast.makeText(this, getString(R.string.pak_fill_paths), Toast.LENGTH_SHORT).show(); return; }
            startOp(() -> {
                boolean ok = PakTool.atlasPack(d, a, p);
                mHandler.post(() -> mStatus.setText(ok ?
                    getString(R.string.pak_complete, a) : getString(R.string.pak_failed)));
            });
        });
        layout.addView(pb);

        mStatus = new TextView(this);
        mStatus.setTextSize(13);
        mStatus.setPadding(0, 16, 0, 0);
        layout.addView(mStatus);

        setContentView(rootScroll);
    }
}

File: app/src/main/java/com/deadcells/modding/AtlasActivity.kt
package com.deadcells.modding

import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.input.rememberTextFieldState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.deadcells.modding.ui.theme.DCMMTTheme
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import top.yukonga.miuix.kmp.basic.Button
import top.yukonga.miuix.kmp.basic.Scaffold
import top.yukonga.miuix.kmp.basic.Text
import top.yukonga.miuix.kmp.basic.TextField
import top.yukonga.miuix.kmp.basic.TopAppBar
import top.yukonga.miuix.kmp.theme.MiuixTheme

class AtlasActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            DCMMTTheme {
                AtlasScreen(
                    onRunOperation = { op ->
                        GlobalScope.launch(Dispatchers.IO) {
                            val ok = op()
                            withContext(Dispatchers.Main) {
                                Toast.makeText(
                                    this@AtlasActivity,
                                    if (ok) getString(R.string.pak_complete, "OK")
                                    else getString(R.string.pak_failed),
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                    }
                )
            }
        }
    }
}

@Composable
fun AtlasScreen(
    onRunOperation: (suspend () -> Boolean) -> Unit
) {
    val atlasState = rememberTextFieldState()
    val outDirState = rememberTextFieldState()
    val pngDirState = rememberTextFieldState()
    val atlasOutState = rememberTextFieldState()
    val pngOutState = rememberTextFieldState()

    Scaffold(
        topBar = {
            TopAppBar(title = stringResource(R.string.atlas_title))
        }
    ) { padding ->
        Column(
            Modifier
                .fillMaxSize()
                .padding(padding)
                .verticalScroll(rememberScrollState())
                .padding(12.dp)
        ) {
            SectionTitle(stringResource(R.string.pak_atlas_unpack))
            TextField(
                state = atlasState,
                label = stringResource(R.string.pak_atlas_unpack_hint),
                useLabelAsPlaceholder = true,
                modifier = Modifier.fillMaxWidth()
            )
            TextField(
                state = outDirState,
                label = stringResource(R.string.pak_unpack_hint_dir),
                useLabelAsPlaceholder = true,
                modifier = Modifier.fillMaxWidth()
            )
            Button(
                onClick = {
                    val ap = atlasState.text.toString()
                    val od = outDirState.text.toString()
                    if (ap.isEmpty() || od.isEmpty()) return@Button
                    onRunOperation { PakTool.atlasUnpack(ap.trim(), od.trim()) }
                },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(stringResource(R.string.pak_atlas_unpack_btn))
            }

            SectionTitle(stringResource(R.string.pak_atlas_pack))
            TextField(
                state = pngDirState,
                label = stringResource(R.string.pak_atlas_pack_hint_dir),
                useLabelAsPlaceholder = true,
                modifier = Modifier.fillMaxWidth()
            )
            TextField(
                state = atlasOutState,
                label = stringResource(R.string.pak_atlas_pack_hint_atlas),
                useLabelAsPlaceholder = true,
                modifier = Modifier.fillMaxWidth()
            )
            TextField(
                state = pngOutState,
                label = stringResource(R.string.pak_atlas_pack_hint_png),
                useLabelAsPlaceholder = true,
                modifier = Modifier.fillMaxWidth()
            )
            Button(
                onClick = {
                    val pd = pngDirState.text.toString()
                    val ao = atlasOutState.text.toString()
                    val po = pngOutState.text.toString()
                    if (pd.isEmpty() || ao.isEmpty() || po.isEmpty()) return@Button
                    onRunOperation { PakTool.atlasPack(pd.trim(), ao.trim(), po.trim()) }
                },
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(stringResource(R.string.pak_atlas_pack_btn))
            }
        }
    }
}
